test/raft/distributed_test.cc

```diff
@@ -31,5 +40,142 @@ requires (typename M::state_t s, typename M::input_t i) {
     requires std::is_same_v<const typename M::state_t, decltype(M::init)>;
 };
 
+using cmd_id_t = utils::UUID;
+
+template <typename State>
+using snapshots_t = std::unordered_map<raft::snapshot_id, State>;
+
+template <PureStateMachine M>
+struct impure_state_machine : public raft::state_machine {
+    typename M::state_t _val;
+    snapshots_t<typename M::state_t>& _snapshots;
+
+    bool _aborted = false;
+
+    // To obtain output from an applied command, the client (`call`)
+    // first allocates a channel in this data structure by calling `allocate`
+    // and makes the returned command ID a part of the command passed to Raft.
+    // When (if) we eventually apply the command, we use the ID to find
+    // the output channel here and push the output to the client waiting
+    // on the other end.
+    // The client is responsible for deallocating the channels, both in the case
+    // where they manage to obtain the output and when they give up.
+    // The client allocates a channel only on its local server; other replicas
+    // of the state machine will therefore not find the ID in their instances
+    // of _output_channels so they just drop the output.
+    using channels_t = std::unordered_map<cmd_id_t, promise<typename M::output_t>>;
+    using channel_handle = typename channels_t::const_iterator;
+    channels_t _output_channels;
+
+    impure_state_machine(snapshots_t<typename M::state_t>& snapshots)
+        : _val(M::init), _snapshots(snapshots) {}
+
+    future<> apply(std::vector<raft::command_cref> cmds) override {
+        for (auto& cref : cmds) {
+            auto is = ser::as_input_stream(cref);
+            auto cmd_id = ser::deserialize(is, boost::type<cmd_id_t>{});
+            auto input = ser::deserialize(is, boost::type<typename M::input_t>{});
+            auto [new_state, output] = M::delta(std::move(_val), std::move(input));
+            _val = std::move(new_state);
```

Comment:
So the pure state machine is so pure that it does not even have its state? It just defines types and transition function?

Reply:
Exactly. The purest you can have.